### ISSUE TITLE
UDP check for service stanza

### ISSUE THREAD

**Issue by gfonseca-tc** [hashicorp/consul#12221]
Please search the existing issues for relevant feature requests, and use the reaction feature (https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to add upvotes to pre-existing requests.

#### Feature Description

Consul support many different check types but not UDP. It could be very useful for all sorts of applications that listen to UDP ports.

#### Use Case(s)

One example: the StatsD receiver, in the OpenTelemetry collector, listens to UDP. Currently there is no way of creating a Consul check against it.

**Comment by Amier3** [comment]
Hey @gfonseca-tc 

Thanks for making this request! We've gotten similar requests in the past for UDP check support, so I appreciate you bringing this back to the forefront 👍

**Comment by fulviodenza** [comment]
I'd love to work on this with someone else from hashicorp if possible @Amier3

**Comment by dhiaayachi** [comment]
Hey @fulviodenza,
Sorry for the late response I was caught on other stuff. Here is some notes that we have about  [the possible entry points](https://github.com/hashicorp/consul/blob/main/docs/service-discovery/health-checks.md) for a check in Consul. 

I hope that can get you started, let us know if you have any questions?

**Comment by fulviodenza** [comment]
Hi @dhiaayachi thank you for the response! I've read the agent/checks/check.go file and It was pretty clear the way to implement UDP check, but I think I'll need help to test it. I think I'll start the development of this feature next week.
Before starting implementation, I'll analyze the changing to do and write down all files and struct that I think will be affected, if this list will be complete, I'll start the implementation. 
I'll write in this thread next few days with this list. Let me know if something doesn't sound good.

**Comment by fulviodenza** [comment]
OK @dhiaayachi, here I am, I just located these files that need to be changed, if there are other files to change tell me:

- [ ] [agent/agent.go](https://github.com/hashicorp/consul/blob/main/agent/agent.go)
- [ ] [agent/check.go](https://github.com/hashicorp/consul/blob/main/agent/check.go)
- [ ] [agent/structs/structs.go](https://github.com/hashicorp/consul/blob/main/agent/structs/structs.go)
- [ ] [agent/structs/check_type.go](https://github.com/hashicorp/consul/blob/main/agent/structs/check_type.go)
- [ ] [agent/config/config.go](https://github.com/hashicorp/consul/blob/main/agent/config/config.go)
- [ ] [api/health.go](https://github.com/hashicorp/consul/blob/main/api/health.go)
- [ ] [proto/pbservice/healthcheck.pb.go](https://github.com/hashicorp/consul/blob/main/proto/pbservice/healthcheck.pb.go)

When I'll finish with each of these files I'll mark it as done in this issue, so we can track up the issue status.
If I missed something, please let me know.

About the last message, I think the information needed for a UDP check is the same as the TCP connection but the TCP field should be UDP with the related IP and port. Isn't it? 
A UDP datagram is composed of port, length and two optional fields (source port and checksum). I think we need just the necessary fields (passed as parameters in the configuration, as shown in this doc file https://www.consul.io/docs/discovery/checks).
to summarise (I leave parameters as they are in TCP example but they don't have to be necessarily that):
```
check = {
  id = "ssh"
  name = "SSH UDP on port xx"
  udp = "localhost:xx"
  interval = "10s"
  timeout = "1s"
}
```
where xx is the port.

**Comment by dhiaayachi** [comment]
> A UDP datagram is composed of port, length and two optional fields (source port and checksum). I think we need just the necessary fields (passed as parameters in the configuration, as shown in this doc file https://www.consul.io/docs/discovery/checks).
to summarise (I leave parameters as they are in TCP example but they don't have to be necessarily that):




I don't think this would work, unlike TCP, opening a socket (connexion) to a UDP endpoint won't give any feedback on whether a process is listening to this address or not. You need to send data to that endpoint and get a reply from the listening process to validate that it's effectively listening to that address (IP + port). The sent payload and the response in this case are application specific and need to be defined as part of the check configuration as we want Consul udp checks to be generics and not for a specific use case.

**Comment by fulviodenza** [comment]
I am missing something, the payload should be the same for each datagram since this should be used just for a health check.
I think there should be just a default payload towards the endpoint and a payload "healthy"/"not healthy" from the endpoint.
Why should the user customize the payload?

**Comment by gfonseca-tc** [comment]
@dhiaayachi, I think this approach will be very problematic. Many applications will not have support for responding the request like you mentioned. The applications we create we could do that, but the way I see it, checking if the port is open should be enough. Something like we do with netcat, something like `nc -z -u <host> <port>`.
I think @fulviodenza approach should cover most of the cases. Maybe the other fields could be optional.

**Comment by dhiaayachi** [comment]
@gfonseca-tc, this mean we are relying on ICMP answers for closed port (this is how netcat detect open ports). It actually detect that the port is closed when it receive an ICMP error from the OS and assume it's open if no response is received. 

This approach is problematic because it will set the health to pass if no packet is received, so this is not reliable specially that a lot of network appliances won't allow ICMP through, so on this case the health-check will always be passing even if the port is closed or unreachable.

That said, I will bring this to the team in our next triage meeting and comment with more details on the approach that we decided for.

**Comment by gfonseca-tc** [comment]
Hey @dhiaayachi , I agree this can be problematic with some network configurations but it's better than nothing and covers lots of cases with no need of changing code in the application. The payload part although is not true, the `-z` option just scans the port, it does not send any data.

**Comment by dhiaayachi** [comment]
Hey @fulviodenza,
sorry for the delay, I finally got the chance to discuss this with the team. For the time being we think it's good enough to have the check use ICMP (the way you designed it). We will add more details about the limitation in the associated documentation and we can add options to specify the payload later.

Let me know if you are still interested in contributing this, I will be out until April 1st but I can be your main contact to review PRs or answer questions when I'm back.

**Comment by nergdron** [comment]
as a side party watching this, I'm glad to see this moving forward. I'm definitely hoping there'll be some future hooks for different protocol / payload types, as we're seeing a lot more big-name UDP based protocols these days (QUIC / http3). we're already using QUIC in production, so I hope the team is thinking forward to those protocol challenges.

**Comment by jkirschner-hashicorp** [comment]
Hi @nergdron,

Can you share with us more information how what you'd hope the check interface might need to look like to support your use cases, like @fulviodenza [did with the ICMP-based basic UDP check](https://github.com/hashicorp/consul/issues/12221#issuecomment-1039348959)?
```
check = {
  id = "ssh"
  name = "SSH UDP on port xx"
  udp = "localhost:xx"
  interval = "10s"
  timeout = "1s"
}
```

**Comment by fulviodenza** [comment]
Hi @dhiaayachi, I'm still interested in working on this task.
I would bring to attention the @jkirschner-hashicorp question. 
Should the use case be the one I proposed above?

**Comment by jkirschner-hashicorp** [comment]
@fulviodenza: my understanding is that @dhiaayachi was suggesting to proceed with the design you originally proposed and @gfonseca-tc helped to clarify: an ICMP-based UDP check. And that, in the future, we _could_ enhance this check to allow optionally specifying a request payload / expected response. Does this make sense?

(@dhiaayachi: please correct me if needed!)

Thanks for offering to contribute this! When you submit the PR, please demonstrate a test case that shows a failing UDP check (so we're all on the same page about the conditions under which failure can be detected).

**Comment by dhiaayachi** [comment]
@fulviodenza Sorry for the delay I was OOO. As @jkirschner-hashicorp said, yes you can proceed with what you initially proposed and you can assign me on PRs when you have something ready to go through the review process.
Thank you for the contribution

**Comment by fulviodenza** [comment]
Sorry for the delay, I came into some implementation trouble writing the module and the tests. Will the client run a server to accept the server's response? Right? If yes, the client and the server will expect some pre-defined messages? Still, if yes, is there a standard or should I choose an arbitrary message accepted?

**Comment by dhiaayachi** [comment]
Hey @fulviodenza 
I'm not sure I understand the question? Can you please define what are you referring to by `client` and `server`? 
My understanding that Consul will open a connection to the service (as a client) and the service is supposed to have a UDP server running and answer to Consul health-check requests. 

I don't think that Consul should expect any specific response from the Service, any answer is enough to set the service as healthy.

**Comment by fulviodenza** [comment]
Hi @dhiaayachi 
I was testing the module and I needed to mock up a UDP server. sorry for my bad explanation. 
Anyway, the answer you gave was pretty clear. Just one more thing,
Consul will expect any response, this means there will be a goroutine in the health-check waiting for some response. this goroutine will be a UDP server or something other?

**Comment by dhiaayachi** [comment]
> Consul will expect any response, this means there will be a goroutine in the health-check waiting for some response. this goroutine will be a UDP server or something other?
No this should happen in the Client goroutine as a state machine in Consul, so the code should go as follow:

- Consul call `net.DialUDP` to the service
- Consul call `_, err = conn.Write("ping")`
- Consul call `_, err = c.Read(buffer)`
- if any of the calls fail the service is marked unhealthy (except of timeout on read, we treat that as success), if all successful service is healthy.
- Consul sleep for the check time and repeat the cycle again
